sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java

```diff
@@ -0,0 +1,980 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet;
+
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.column.Dictionary;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.IntLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.DecimalLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.TimestampLogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+
+import org.apache.spark.sql.catalyst.util.DateTimeUtils;
+import org.apache.spark.sql.catalyst.util.RebaseDateTime;
+import org.apache.spark.sql.execution.datasources.DataSourceUtils;
+import org.apache.spark.sql.execution.datasources.SchemaColumnConvertNotSupportedException;
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.DecimalType;
+
+import java.math.BigInteger;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+
+public class ParquetVectorUpdaterFactory {
+  private static final ZoneId UTC = ZoneOffset.UTC;
+
+  private final LogicalTypeAnnotation logicalTypeAnnotation;
+  // The timezone conversion to apply to int96 timestamps. Null if no conversion.
+  private final ZoneId convertTz;
+  private final String datetimeRebaseMode;
+  private final String int96RebaseMode;
+
+  ParquetVectorUpdaterFactory(
+      LogicalTypeAnnotation logicalTypeAnnotation,
+      ZoneId convertTz,
+      String datetimeRebaseMode,
+      String int96RebaseMode) {
+    this.logicalTypeAnnotation = logicalTypeAnnotation;
+    this.convertTz = convertTz;
+    this.datetimeRebaseMode = datetimeRebaseMode;
+    this.int96RebaseMode = int96RebaseMode;
+  }
+
+  public ParquetVectorUpdater getUpdater(ColumnDescriptor descriptor, DataType sparkType) {
+    PrimitiveType.PrimitiveTypeName typeName = descriptor.getPrimitiveType().getPrimitiveTypeName();
+
+    switch (typeName) {
+      case BOOLEAN:
+        if (sparkType == DataTypes.BooleanType) {
+          return new BooleanUpdater();
+        }
+        throw constructConvertNotSupportedException(descriptor, sparkType);
+
+      case INT32:
+        if (sparkType == DataTypes.IntegerType || canReadAsIntDecimal(descriptor, sparkType)) {
+          return new IntegerUpdater();
+        } else if (sparkType == DataTypes.LongType) {
+          // In `ParquetToSparkSchemaConverter`, we map parquet UINT32 to our LongType.
+          // For unsigned int32, it stores as plain signed int32 in Parquet when dictionary
+          // fallbacks. We read them as long values.
+          return new UnsignedIntegerUpdater();
```

Comment:
shall we check if the logical type is UINT32?

Reply:
I just directly translated the existing logic to here. It seems this has been discussed previously: https://github.com/apache/spark/pull/31921/files#diff-09a6513eee0907b1a6610471aa4ebd6af2bef01bf4b5c082dc8c9ad1d6b030d9R583